master/buildbot/process/buildrequestdistributor.py

```diff
@@ -422,13 +422,35 @@ def claimBuildRequests(self, breqs):
 
     def removeBuildRequest(self, breq):
         # reset the checkMerges in case the breq still in the master cache
+
+        def remove_dict(brdicts, breq):
+            old_len_brdicts = len(brdicts)
+            brdicts = [brdict for brdict in brdicts if brdict['brid'] != breq.brdict['brid']]
+            is_removed = old_len_brdicts != len(brdicts)
+            return {
+                'brdicts': brdicts,
+                'is_removed': is_removed
+            }
+
         breq.checkMerges = True
         breq.retries = 0
         breq.hasBeenMerged = False
-        if self.unclaimedBrdicts and breq.brdict and breq.brdict in self.unclaimedBrdicts:
-            self.unclaimedBrdicts.remove(breq.brdict)
-        if self.resumeBrdicts and breq.brdict and breq.brdict in self.resumeBrdicts:
-            self.resumeBrdicts.remove(breq.brdict)
+        is_removed = False
+
+        if self.unclaimedBrdicts and breq.brdict:
+            unclaimed_data = remove_dict(self.unclaimedBrdicts, breq)
+            self.unclaimedBrdicts = unclaimed_data['brdicts']
+            is_removed |= unclaimed_data['is_removed']
+
+        if self.resumeBrdicts and breq.brdict:
+            resume_data = remove_dict(self.resumeBrdicts, breq)
+            self.resumeBrdicts = resume_data['brdicts']
+            is_removed |= resume_data['is_removed']
+
+        msg = "removeBuildRequest does not remove anything. breq.brid: %s" % breq.brdict['brid']
+        if not is_removed:
+            klog.err_json(msg)
```

Comment:
That looks more complicated than necessary :)

I thought about something like this (unless you see something that wouldn't work):

```python
def removeBuildRequest(self, breq):
    # reset the checkMerges in case the breq still in the master cache
    def remove_dict(brdicts):
        for brdict in brdicts:
            if brdict['brid'] == breq.brdict['brid']:
                brdicts.remove(brdict)
                return True
        return False

    if not breq.brdict:
        pass # log some error here
    
    removed = remove_dict(self.unclaimedBrdicts) or remove_dict(self.resumeBrdicts)
    if removed:
        breq.checkMerges = True
        breq.retries = 0
        breq.hasBeenMerged = False
        self.breqCache.remove(breq.id)
    else:
        klog.err_json(
            "removeBuildRequest does not remove anything. breq.brid: %s" % \
            breq.brdict['brid']
        )
```

Reply:
list comprehension removes all dicts with the ['brid'] so we don't leave a possible mess in memory. In your case we remove only one dict from list. @Mowinski What do you think?

Reply:
They shouldn't show up multiple times in there, but you can do it list this to feel safer:

```python
def remove_dict(brdicts):
    removed = False
    for brdict in brdicts:
        if brdict['brid'] == breq.brdict['brid']:
            brdicts.remove(brdict)
            removed = True
    return removed
```

Reply:
I'm not a huge fan of replacing the entire list with a comprehension because I don't know if anyone else has a hard reference to the old list.

Reply:
As I said below, I am not a fan of inner function in this place :) and I agree with Diogo, whats more, as I see we modify the property of class so even we do not need an inner function to do it and a private method should be enough to do it.

Reply:
OK, I've used Diogo's code